Add EQ-Bench as per #1459

pbevan1 opened the pull request:
Implemented EQ-Bench as per #1459

This uses the latest v2 data and scoring system which aligns with the leaderboard at https://eqbench.com/index.html

@djstrong feel free to test this as you mentioned on the issue. I tested with a couple OpenAI models and get pretty close matches to the online leaderboard:

gpt-4-0125-preview: tested=82.73, leaderboard=83.87
gpt-3.5-turbo-0125: tested=67.65, leaderboard=64.97

This is my first task implementation here, and I've tried to follow the docs for everything, but please let me know if anything is incorrect and I'll fix it.

haileyschoelkopf commented:
Thanks for the PR! It'd be nice if we could get @sqrkl to take a look at this.

I'll get this run on a couple open source models and report back!

Actually, tried to give this a spin and got
```
 AttributeError'NoneType' object has no attribute 'items'
 ```
are there some updates to your `process_results` function that aren't pushed here?

sqrkl commented:
Great! Thank you for doing this @pbevan1 

I'll check out the error and run it through a few models to validate tomorrow.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/EleutherAI/lm-evaluation-harness?pullRequest=1511) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/EleutherAI/lm-evaluation-harness?pullRequest=1511) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: pbevan1<br/>:x: Peter Bevan<br/><hr/>**Peter Bevan** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/EleutherAI/lm-evaluation-harness?pullRequest=1511) it.</sub>

pbevan1 commented:
Ah, this error is from not properly handling cases where the answer isn't parsable, and this didn't occur with the OpenAI models so I didn't spot it. I fixed it to skip these cases as per the paper:

> The final score is the average of scores for all answers that were parsable. To focus on assessing
emotional intelligence rather than formatting proficiency, only parsable answers are considered in the
final score average. In practice, most models produced near to 60 parsable answers. If fewer than 50
out of 60 questions were parsable, the test is considered a fail.

The only thing that is not implemented is the failure if fewer than 50 answers are parsable. I'm not sure on the best way to go about implementing this since the aggregation is done in the main harness code..

Tested on a couple local models:

vince62s/phi-2-psy: tested=55.71, leaderboard=56.44
mistralai/mistral-7b-instruct-v0.1: tested=45.81, leaderboard=52.15

sqrkl commented:
The current logic in EQ-Bench v2 is that 83% of answers have to be parseable (it has 171 questions over v1's 60). 

The issue with not including a fail mode is that if a model produces just 1 parseable answer, and that answer happens to score well by chance, its score will be determined by that one question.

The two approaches I can see here are:

1. create a custom aggregation function that sets the score to zero or some fail state (preferably with a message why it failed). this mirrors the eq-bench logic
2. use mean aggregation for the full number (171) questions, considering unparseable answers as 0

Option 2 will produce lower scores for some models compared to the eq-bench pipeline, which isn't ideal. Not sure how feasible option 1 is; I know we can create the custom aggregation function, but not sure about if tests can be failed w/ explanatory output.

djstrong commented:
I like the second option. I am working on smaller models which may not generate proper formatting for more than 83% of cases and still I want to rank them.

sqrkl commented:
> I like the second option. I am working on smaller models which may not generate proper formatting for more than 83% of cases and still I want to rank them.

That's a good point. Well I would be fine with that (and it simplifies things). Would be good to make a note of the differences to the eq-bench pipeline on the task readme. Specifically:

- no retries in lm-eval version (eq-bench pipeline retries with successively higher temps if it encounters unparseable answers)
- explain the difference in score aggregation with unparseable answers

Some changes I would suggest to the .yaml task definition:

```
generation_kwargs:
  do_sample: false
  temperature: 0.0
  max_gen_toks: 80
```

@pbevan1 how does that sound?

sqrkl commented:
Oh also, it would be good to return n_parseable as a metric.

pbevan1 commented:
On it

pbevan1 commented:
Ok, this should be working as agreed now:

- Unparseable answers now return a score of 0.
- Added the gen kwargs
- Added notes in README to flag the differences between this and original implementation
- Added percent_parseable metric (went with percent as seems more interpretable, let me know if that's ok)

We get results like this:

![image](https://github.com/EleutherAI/lm-evaluation-harness/assets/69802263/8b592658-8939-41d4-9ef8-227058d6a8eb)

pbevan1 commented:
@sqrkl Done. Seems like we should be good to merge if you're also happy @haileyschoelkopf

haileyschoelkopf commented:
Nice, awesome!

If we did want to implement the "fail" mode, it'd be possible:
- have your process_results function return -1 or some sentinel for the "eqbench" key
- have an aggregation function which checks how many -1 values there are, and returns 0 if there are too many, or else excludes these -1 values and calculates the EQ bench score
	
but otherwise I'm good with it since @sqrkl is.
